### PR TITLE
AP-110: focus trap for interactive-opened dialogs

### PR DESCRIPTION
**Pull request by scytacki**
## Summary

- Integrates `@concord-consortium/accessibility-tools` 0.2.0-pre.2 and wraps interactive-opened dialogs in a focus trap (AP-110): keyboard users can reach the close button, Tab descends into the iframe, and Tab/Shift+Tab cycles back to the close button without escaping the modal. Includes Safari cross-origin Tab-out fix from the library.
- Adds a focus-debug overlay enabled via `localStorage.setItem("debug", "focus")` that logs focusin/focusout, Tab/Enter/Escape keydowns, and window blur/focus — built to triage Safari's cross-origin focus race that surfaced during AP-110.

### A note on the focus-debug overlay

This overlay is throwaway debug tooling that overlaps with accessibility-tools, kept here for now but intended to be folded upstream when there's time:

- **Overlap:** the library's `AccessibilityDebugSidebar` (`/debug` entry, also a hosted bookmarklet) already has richer focus-tracker, focus-history, and keyboard-log panels that duplicate most of what this overlay shows.
- **What's different:** this overlay detects focus crossing into a cross-origin iframe — the library's focus stream listens only to `document` `focusin`, which never fires when the parent yields focus to an iframe. Here we watch window `blur`/`focus` (logging "iframe gained focus" when `document.activeElement` becomes the iframe) and poll `activeElement`. That iframe signal is the reason this overlay was written — it tracked down a tricky Safari focus bug, where the library's bookmarklet path is also awkward to use.
- **How to merge:** move the window blur/focus + `activeElement` poll upstream into accessibility-tools' focus stream as a synthetic "focus entered iframe" event. Then AP can delete this overlay and mount the library sidebar directly (gated by `DEBUG_FOCUS`) — which works in Safari too, since it's the imported-component path rather than a bookmarklet.

Jira: https://concord-consortium.atlassian.net/browse/AP-110
Design: [specs/AP-110-dialog-overlay-focus-trap.md](specs/AP-110-dialog-overlay-focus-trap.md)

Two links for testing:
- https://activity-player.concord.org/branch/dialog-overlay-focus-trap/index.html?activity=sample-activity-all-question-interactives - go to page 9 and click the Make Drawing button (either with mouse or keyboard), and notice the hint telling you to tab into to the interactive. The drawing tool used in this version of the image question has no keyboard navigation, so you won't see any focus rings until you tab to the text area or buttons in the dialog
- https://activity-player.concord.org/branch/dialog-overlay-focus-trap/index.html?activity=sample-activity-all-question-interactives&override.qi=toolbar-accessibility - same activity but now with a version of the image question that has keyboard navigation.

## Test plan

- [ ] Open the image-question dialog in Chrome — Tab forward from the close button descends into the iframe; Shift+Tab from the close button shows the after-sentinel landing hint at the iframe's top-left
- [ ] Same flow in Firefox (macOS)
- [ ] Same flow in Safari with "Keyboard navigation" enabled in System Settings — verify the Tab-out-of-iframe path lands on the before-sentinel hint (not stranded on the after-sentinel)
- [ ] Close the dialog (button click, Escape on close button, scrim click, in-iframe Cancel/Done) — focus restores to the originating iframe
- [ ] `localStorage.setItem("debug", "focus")` + reload: the focus debug overlay renders above the dialog scrim and logs Tab/focus events
- [ ] Unit tests pass: `dialog-overlay`, `iframe-runtime`, `managed-interactive`, `use-composite-ref`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/components/activity-page/managed-interactive/dialog-overlay.tsx [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+  return (
+    <Modal
+      isOpen={true}
+      appElement={getModalContainer()}
+      shouldFocusAfterRender={false}
+      shouldReturnFocusAfterClose={false}
+      shouldCloseOnEsc={false}
+      onRequestClose={safeOnClose}
+    >
```
[the remark this answers is not in the document]
Comment (scytacki) [reply src/components/activity-page/managed-interactive/dialog-overlay.tsx:134]: Fixed in af70d27. \`onRequestClose\` is now \`notCloseable ? undefined : safeOnClose\`, restoring master's behavior where overlay clicks can't dismiss a non-closeable dialog. Added \`contentLabel={title || "Dialog"}\` as well.

File: src/components/activity-page/managed-interactive/managed-interactive.tsx [hunk 2]
```diff
@@ -428,10 +441,15 @@ export const ManagedInteractive: React.ForwardRefExoticComponent<IProps> = forwa
...
-              activeDialog &&
-              <Modal isOpen={true} appElement={getModalContainer()} onRequestClose={activeDialog.notCloseable ? undefined : handleCloseDialog}>
-                { interactiveIframeRuntime }
-              </Modal>
+              activeDialog && (loadingAnswer || loadingLegacyLinkedInteractiveState
+                ? "Loading..."
+                : <DialogOverlay
+                    url={activeDialog.url || (embeddable.url_fragment ? url + embeddable.url_fragment : url)}
+                    notCloseable={activeDialog.notCloseable}
+                    onClose={handleCloseDialog}
```
[the remark this answers is not in the document]
Comment (scytacki) [reply src/components/activity-page/managed-interactive/managed-interactive.tsx:450]: Fixed in af70d27 — now passes the \`iframeUrl\` already computed above.

File: src/components/activity-page/managed-interactive/iframe-runtime.tsx [hunk 3]
```diff
@@ -514,11 +521,21 @@ export const IframeRuntime: React.ForwardRefExoticComponent<IProps> = forwardRef
...
+        </span>
+      </span>
       <iframe
         className={locked ? "iframe-runtime-locked" : ""}
         tabIndex={locked ? -1 : 0}
         key={`${id}-${reloadCount}`}
-        ref={iframeRef}
+        ref={composeRefs(iframeRef, externalIframeRef)}
         src={applyOverrides(url)}
         id={id}
```
[the remark this answers is not in the document]
Comment (scytacki) [reply src/components/activity-page/managed-interactive/iframe-runtime.tsx:543]: Fixed in af70d27. Folded the memoization into a new \`useCompositeRef\` hook (renamed from \`compose-refs\` → \`use-composite-ref\`), so the iframe gets a stable composed ref and callers can't forget to memoize. \`composeRefs\` stays as the pure helper.

File: specs/AP-110-dialog-overlay-focus-trap.md [hunk 4]
```diff
@@ -0,0 +1,272 @@
+# AP-110: Focus Trap for the Dialog Overlay (non-cooperating)
+
+**Jira**: https://concord-consortium.atlassian.net/browse/AP-110
+
+**Parent design**: [AP-108-focus-traps-cross-origin-interactives.md](AP-108-focus-traps-cross-origin-interactives.md) — AP-110 is the first piece of AP-108 work to land in Activity Player.
+
+**Scope**: Activity Player only. Integrate `@concord-consortium/accessibility-tools@0.2.0-pre.0` and apply it to the **dialog** modal that an interactive opens via `showModal({ type: "dialog" })`. Non-cooperating only — the interactive does not yet speak the focus-coordination protocol. The image-question interactive is the verification target.
```
[the remark this answers is not in the document]
Comment (scytacki) [reply specs/AP-110-dialog-overlay-focus-trap.md]: Fixed in af70d27 — updated the spec to \`0.2.0-pre.2\` to match \`package.json\`.

File: src/components/activity-page/managed-interactive/dialog-overlay.tsx [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+// `@concord-consortium/accessibility-tools/hooks` is resolved via a
+// moduleNameMapper entry in package.json that points at the dist CJS file.
+// When the project upgrades to Jest 28+, the mapper can be removed.
+import { useFocusTrap, useIframeSlot, FocusTrapStrategy } from "@concord-consortium/accessibility-tools/hooks";
+import { IframeRuntime, IframeRuntimeImperativeAPI } from "./iframe-runtime";
+import "./dialog-overlay.scss";
+
+const getModalContainer = (): HTMLElement => {
+  return document.getElementById("app") || document.body;
+};
```
[the remark this answers is not in the document]
Comment (scytacki) [reply src/components/activity-page/managed-interactive/dialog-overlay.tsx:13]: Leaving this as-is. The \`document.body\` fallback is pre-existing (same \`getModalContainer\` on master) and never triggers in production, since AP always renders into \`#app\` — so \`appElement\` is \`#app\` in practice and the modal content is not hidden. Tightening it to require \`#app\` would only affect the missing-element edge case (e.g. tests), with no real-world benefit.

File: src/index.tsx [hunk 6]
```diff
@@ -1,6 +1,8 @@
 import React from "react";
 import ReactDOM from "react-dom";
 import { App } from "./components/app";
+import { FocusDebugOverlay } from "./components/focus-debug-overlay";
+import { DEBUG_FOCUS } from "./lib/debug";
```
[the remark this answers is not in the document]
Comment (scytacki) [reply src/index.tsx:5]: I don't think this is worth it, the code is small enough that it is OK to import it all of the time.

File: src/components/focus-debug-overlay.tsx [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+  // Portal directly to <body> so the overlay competes for stacking with
+  // react-modal's own portal (also a body-level sibling). Rendering inside
+  // <App /> traps the overlay in #app's stacking context, which the dialog's
+  // body-level portal then paints over.
+  return ReactDOM.createPortal(
+    <div className="focus-debug-overlay" tabIndex={-1}>
+      <div className="focus-debug-overlay__header">
+        <span>focus debug</span>
+        <button type="button" onMouseDown={e => e.preventDefault()} onClick={clear}>clear</button>
+      </div>
```
[the remark this answers is not in the document]
Comment (scytacki) [reply src/components/focus-debug-overlay.tsx:134]: This is true, but since this code has mostly served its purpose, it doesn't seem worth modifying until we have need to do so.

File: src/components/focus-debug-overlay.tsx [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    };
+    const onFocusOut = (e: FocusEvent) => {
+      const target = e.target as Element | null;
+      push("focusout", describe(target));
+    };
+    const onKey = (e: KeyboardEvent) => {
+      if (e.key !== "Tab" && e.key !== "Enter" && e.key !== "Escape") return;
+      const mod = e.shiftKey ? "Shift+" : "";
+      push("key", `${mod}${e.key} @ ${describe(e.target as Element)}`);
+    };
```
[the remark this answers is not in the document]
Comment (scytacki) [reply src/components/focus-debug-overlay.tsx:93]: Fixed in 926b493. Widened \`describe()\`/\`path()\` to accept \`EventTarget | null\` and narrow with \`instanceof Element\` (returning "(none)" otherwise), which also covers the same cast in the focusin/focusout handlers and drops the unsafe \`as Element\` casts.

**Review by kswenson — APPROVED** [review]
👍 Looks great -- the following review, developed in conjunction with Claude Code 🤖, includes the results of an accesslint accessibility audit, along with some suggested improvements to be addressed or deferred at your discretion.

---

## PR Review Summary

**Changes:** 17 files, +1184 / −55

### What it does

Integrates `@concord-consortium/accessibility-tools@0.2.0-pre.2` and wraps the dialog modal an interactive opens (`showModal({ type: "dialog" })`) in a keyboard focus trap (AP-110, the first slice of AP-108). A keyboard user can now reach a host-rendered Close button, Tab forward into the iframe, and Tab/Shift+Tab cycle back without escaping the modal — including the Safari cross-origin Tab-out fix from the library. Also ships a dev-only focus-debug overlay behind `localStorage.setItem("debug","focus")`.

### The approach

- **New `DialogOverlay`** owns the `<Modal>`, header/close button, and the trap (`useFocusTrap` + `useIframeSlot`) — keeping this out of the already-large `managed-interactive.tsx`. react-modal's auto-focus/return-focus/esc are all disabled so the trap is the single owner; a `closedRef` one-shot guard funnels every dismiss path through `safeOnClose` to avoid a double-close on unmount.
- **`iframe-runtime.tsx`** renders `[before-sentinel][iframe][after-sentinel]` unconditionally (inert when refs aren't wired), gains optional ref props and a `getIframeElement()` imperative method, and uses a new `useCompositeRef` helper to merge internal + external iframe refs without churning the node each render.
- **`managed-interactive.tsx`** factors the long `IframeRuntime` prop list into a single shared `iframeRuntimeProps` object, swaps the inline `<Modal>` for `<DialogOverlay>`, and adds a coarse focus-restore effect (`getIframeElement()?.focus()`) on close.

### Assessment

Clean, focused, and unusually well-commented — every non-obvious decision (ref-callback engagement vs. a mount effect, the forward-declared `trapRef` to break the slot↔trap construction cycle, the double-close guard, the body-level portal for the debug overlay) carries a rationale. CI is fully green: Jest, S3 deploy, all three Cypress shards, and the 96-test cloud run. Good unit coverage for the new units (`dialog-overlay`, `iframe-runtime` sentinels, and a thorough `use-composite-ref`).

I traced two things worth recording as verified-good:

- **The sentinel ref "double-wiring"** in `dialog-overlay.tsx` (the `beforeSentinelRef`/`afterSentinelRef` useRefs passed into `useIframeSlot` while `*SentinelProps.ref` is what attaches to the DOM) is **not** dead code. `*SentinelProps.ref` is the library's `setBeforeSentinel` callback ref, whose body writes `beforeSentinelRef.current = node`, which the library reads back via `getBeforeSentinel`. The passed-in refs are load-bearing. Your TODO there is a valid *upstream* API-ergonomics note, not an AP bug.
- **The `strategy` `useMemo`** (lines 90–103) is robust to React discarding the cache. `useFocusTrap`'s only reaction to a new `strategy` identity is `controller.setStrategy(s)`, whose entire body is `this.strategy = s` — a reference swap, no listener churn / re-entry / focus reset. The memo is a pure referential-stability optimization that degrades gracefully.

### Notes / suggestions (non-blocking)

1. **Spec/test gap (low).** The spec's Testing section promised `managed-interactive.test.tsx` would cover the dialog path rendering `DialogOverlay` and the focus-restoration fallback, but that file wasn't touched. `DialogOverlay` is tested in isolation, so the gap is just the *wiring* — the new `previousActiveDialogRef` focus-restore effect and the `loadingAnswer`-guarded dialog branch are untested. Worth a follow-up.
2. **Debug overlay in the production bundle (low — deliberate).** `focus-debug-overlay` (151 LOC + scss) is statically imported in `index.tsx`, so it ships for all users even though it only renders behind the localStorage flag. You've documented the plan to fold it upstream; flagging only for completeness.
3. **Optional refactor (nitpick).** The one-shot guard in `dialog-overlay.tsx:38–43` (`closedRef` + a `safeOnClose` that early-returns once fired) is a reusable pattern — it could become a small `useCallbackOnce(fn, deps)` hook alongside `useCompositeRef`. Purely a suggestion; the inline version reads fine.

### Accessibility audit (accesslint)

Ran a full WCAG audit on the AP-side markup/ARIA/contrast (the focus-trap mechanics inside `accessibility-tools` were out of scope). The dialog's core semantics are sound — `role="dialog"`, `aria-modal="true"`, and background app-inerting all come from react-modal automatically, and focus-on-open / focus-restoration-on-close are wired correctly. Two findings have real end-user impact with small fixes; none is blocking — address or defer at your discretion.

- **S-1 (Serious, WCAG 4.1.2) — dialog accessible name is always the generic "Dialog".** `IShowDialog` has no `title` field and `managed-interactive` passes no `title` to `DialogOverlay`, so `contentLabel` always falls back to `"Dialog"` and the visible header `<span className="dialog-overlay__title">` is always empty. A screen-reader user opening (e.g.) the image-question dialog hears just "Dialog"; indistinguishable with multiple interactives on a page. *Suggested fix:* pass the already-computed `iframeRuntimeProps.iframeTitle` ("Question N … content") through as `title`/`contentLabel` (and either render it visibly in the header or drop the empty span). This also makes the bare "Close" button name acceptable by supplying context.
- **S-2 (Serious, WCAG 1.4.11) — orange focus ring / hint border fails non-text contrast.** `dialog-overlay.scss:22` (`outline: 2px solid orange` on the close button's focus ring) and `iframe-runtime.scss` (`border: 2px solid orange` on the landing hint) both use `orange` (#ffa500) = **1.97:1 on white**, below the 3:1 minimum. The close-button ring is the primary keyboard focus indicator the trap lands on, so this undercuts the feature's own purpose. The `$cc-orange` token (#ff8415) is only 2.45:1 and also fails; something like `#c2410c` (5.18:1) clears it. *Suggested fix:* swap to a 3:1-compliant token in both SCSS files.
- **M-1 (Moderate) — sentinel hint wording.** Both sentinels and the library `enterLabel` use the identical "Press Tab to enter the interactive"; on the backward/wrap landing (after-sentinel) that's directionally misleading. Best resolved during the manual screen-reader pass the spec already mandates.
- **M-2 (Moderate, WCAG 2.1.2 Level A) — `notCloseable` keyboard hard-lock.** No close button, no Escape handler, overlay click is a no-op, so a keyboard-only user has no host escape and depends entirely on an in-iframe dismiss control. This is **pre-existing**, explicitly acknowledged and deferred in the spec (AP-108 follow-up) — flagged for tracking, not as a regression in this PR.
- **Passes:** dialog role + `aria-modal` + app-inerting, single-owner focus-on-open, coarse focus restoration on close, dialog container `tabIndex={-1}`, and the dev-only debug overlay's color contrast (all text 5.7–13.3:1).
